[WebAuthn] Registrations do not return transports

pascoej opened the pull request:
#### 44f96cd7af27e9fa00a75f04def262e4b4173af2
<pre>
[WebAuthn] Registrations do not return transports
<a href="https://bugs.webkit.org/show_bug.cgi?id=272116">https://bugs.webkit.org/show_bug.cgi?id=272116</a>
<a href="https://rdar.apple.com/125794592">rdar://125794592</a>

Reviewed by Brent Fulgham.

This value is not currently passed along in the modern flow. This patch fixes that.

* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.cpp:
(WebCore::convertStringToAuthenticatorTransport):
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.h:
* Source/WebCore/Modules/webauthn/fido/DeviceResponseConverter.cpp:
(fido::convertStringToAuthenticatorTransport): Deleted.
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::toTransports):
(WebKit::WebAuthenticatorCoordinatorProxy::performRequest):

Canonical link: <a href="https://commits.webkit.org/277123@main">https://commits.webkit.org/277123@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8677accd7fafb147c277593e6284f82847eb2d1f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46492 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25648 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49096 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49168 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42533 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48799 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30008 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23111 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37934 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47070 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22671 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40091 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19187 "Found 1 new API test failure: /WPE/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/autoplay-policy (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20044 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41199 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4537 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42864 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41575 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51009 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21498 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17934 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45184 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22789 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44125 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10329 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23201 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22492 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -453,6 +453,16 @@ static inline bool isCrossPlatformRequest(const Vector<AuthenticatorTransport>&
+inline static Vector<AuthenticatorTransport> toTransports(NSArray<ASAuthorizationSecurityKeyPublicKeyCredentialDescriptorTransport> *asTransports)
+{
+    Vector<AuthenticatorTransport> transports;
+    for (ASAuthorizationSecurityKeyPublicKeyCredentialDescriptorTransport asTransport : asTransports) {
```
Using auto here results in
```
/Volumes/webkit/OpenSource/Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:459:10: error: 'auto' deduced as 'id' in declaration of 'asTransport' [-Werror,-Wauto-var-id]
  459 |     for (auto asTransport : asTransports) {
```
(or auto*)
```
/Volumes/webkit/OpenSource/Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:459:10: error: 'auto' deduced as 'id' in declaration of 'asTransport' [-Werror,-Wauto-var-id]
  459 |     for (auto* asTransport : asTransports) {
```

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -453,6 +453,16 @@ static inline bool isCrossPlatformRequest(const Vector<AuthenticatorTransport>&
 
+inline static Vector<AuthenticatorTransport> toTransports(NSArray<ASAuthorizationSecurityKeyPublicKeyCredentialDescriptorTransport> *asTransports)
+{
+    Vector<AuthenticatorTransport> transports;
```
We can steal a bit from WTF::Vector's various `toVector` implementations to make sure we are doing this efficiently:

`transports.reserveInitialCapacity(asTransports.count);`

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -453,6 +453,16 @@ static inline bool isCrossPlatformRequest(const Vector<AuthenticatorTransport>&
     }
 }
 
+inline static Vector<AuthenticatorTransport> toTransports(NSArray<ASAuthorizationSecurityKeyPublicKeyCredentialDescriptorTransport> *asTransports)
```
Why is this `static` function `inline`?

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -453,6 +453,16 @@ static inline bool isCrossPlatformRequest(const Vector<AuthenticatorTransport>&
+    for (ASAuthorizationSecurityKeyPublicKeyCredentialDescriptorTransport asTransport : asTransports) {
+        if (auto transport = convertStringToAuthenticatorTransport(asTransport))
+            transports.append(*transport);
+    }
```
`transports.shrinkToFit();`

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -453,6 +453,16 @@ static inline bool isCrossPlatformRequest(const Vector<AuthenticatorTransport>&
+    Vector<AuthenticatorTransport> transports;
+    for (ASAuthorizationSecurityKeyPublicKeyCredentialDescriptorTransport asTransport : asTransports) {
+        if (auto transport = convertStringToAuthenticatorTransport(asTransport))
+            transports.append(*transport);
```
`   transports.unsafeAppendWithoutCapacityCheck(*transport);`

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -453,6 +453,16 @@ static inline bool isCrossPlatformRequest(const Vector<AuthenticatorTransport>&
+    }
+    return transports;
+}
+
```
In fact, you might be able to just use the `wtf/VectorCocoa.h` implementation of `toVector(NSArray)` by writing an appropriate `makeVectorElement` function:

```
//    std::optional<VectorElementType> makeVectorElement(const VectorElementType*, id arrayElement);
std::optional<AuthenticatorTransport> makeVectorElement(const AuthenticatorTransport*, ASAuthorizationSecurityKeyPublicKeyCredentialDescriptorTransport *arrayElement) {
    return convertStringToAuthenticatorTransport(asTransport));
}

Then below you can just say:

`response.transports = toVector(credential.get().transports);`

And get rid of the whole `toTransports()` method.

brentfulgham replied:
(You probably need to add an include for `wtf/VectorCocoa.h`)

brentfulgham reviewed (approved):
This is fine as-is, but I think it would be *better* if you adopted the canonical `toVector` logic from WTF.

pascoej committed:
[WebAuthn] Registrations do not return transports
https://bugs.webkit.org/show_bug.cgi?id=272116
rdar://125794592

Reviewed by Brent Fulgham.

This value is not currently passed along in the modern flow. This patch fixes that.

* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.cpp:
(WebCore::convertStringToAuthenticatorTransport):
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.h:
* Source/WebCore/Modules/webauthn/fido/DeviceResponseConverter.cpp:
(fido::convertStringToAuthenticatorTransport): Deleted.
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::toTransports):
(WebKit::WebAuthenticatorCoordinatorProxy::performRequest):

Canonical link: https://commits.webkit.org/277123@main
